teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,410 @@
+
+### Transferring Artifacts
+Until we have an Artifact API, when a `Task` or a `StepAction` needs to consume `Artifacts`, it can only do so via `params`. The user of the `Task`/`StepAction` needs to provide it the artifact provenance data. The `Task`/`StepAction` consuming the artifacts will only get the associated metadata which they need to use to fetch the raw data. When passing artifacts between `Tasks/Steps`, only output artifacts of a previous `Task/Step` can be passed as inputs to the consuming `Task/Step`. Because of the structure of [artifact provenance data](#values), we propose passing the artifact provenance data as a json string. **Note** that for large number of artifacts, this string has the potential of getting large enough to cause the `TaskRun` to exceed the `etcd` limit. However, that is true for `artifacts` in general and in the future, we will require some sort of `larger artifacts` mechanism. For, now, this is out of scope.
+
+#### Between Steps
+
+When passing artifacts between `Steps`, we propose the following syntax: `$(steps.<step-name>.outputs.<artifact-name>)` which resolves to the [values](#values) of the associated `output` artifact. If the artifact was not named, the replacement syntax would look like: `$(steps.<step-name>.outputs)`.
+
+`StepActions` can request a `param` to consume artifact provenance data as follows:
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepActions
+metadata:
+  name: step-action-consuming-artifacts
+  description: |
+    This stepaction produces an `output`` artifact called `images`.
+spec:
+  params:
+    - name: image-artifact-provenance
+  env:
+    - name: image-artifact-provenance
+      value: $(params.image-artifact-provenance)
+  image: jq
+  script: |
+    echo ${image-artifact-provenance} | jq
+```
+
+In the above case, the `Task` can pass in the artifacts from a previous `Step` as follows:
+
+```yaml
+apiVersion: v1
+kind: Task
+metadata:
+  name: task-handling-artifacts
+spec:
+  steps:
+    - name: step-producing-artifacts # when a step produces artifacts, it should be named so that it can be referenced.
+      image: bash
+      script: |
+       echo -n "{\"outputs\":[{\"name\":\"images\",\"values\":[{\"url\":\"some-uri\",\"digest\": {\"sha256\": \"some-digest\"}}]}]}" | tee $(step.artifacts.path) 
+    - name: step-consuming-artifacts
+      ref:
+        name: step-action-consuming-artifacts 
+      params:
+        - name: image-artifact-provenance
+          value: $(steps.step-producing-artifacts.outputs.images) # or $(steps.step-producing-artifacts.outputs ) if the artifact catrgory was not named.
```

Comment:
If we're going to support this, ideally it'd be great to reference the values without needing to introduce jq wrangling.

e.g.

Value | Description
---|---
$(steps.foo.outputs) | Full json
$(steps.foo.outputs.digest) | Single string digest, or error if ambiguous
$(steps.foo.outputs.digests) | Array of `alg:hash` strings
$(steps.foo.outputs.urls[0]) | String of first url, or error if index doesn't exist